cmd/plugins/memory-qos/main.go

```diff
@@ -0,0 +1,304 @@
+	SwapLimitRatio float32
+}
+
+const (
+	annotationSuffix = ".memory-qos.nri.io"
+)
+
+var (
+	log *logrus.Logger
+)
+
+// Configure handles connecting to container runtime's NRI server.
+func (p *plugin) Configure(ctx context.Context, config, runtime, version string) (stub.EventMask, error) {
+	log.Infof("Connected to %s %s...", runtime, version)
+	if config != "" {
+		log.Debugf("loading configuration from NRI server")
+		if err := p.setConfig([]byte(config)); err != nil {
+			return 0, err
+		}
+		return 0, nil
+	}
+	return 0, nil
+}
+
+// onClose handles losing connection to container runtime.
+func (p *plugin) onClose() {
+	log.Infof("Connection to the runtime lost, exiting...")
+	os.Exit(0)
+}
+
+// setConfig applies new plugin configuration.
+func (p *plugin) setConfig(config []byte) error {
+	log.Tracef("setConfig: parsing\n---8<---\n%s\n--->8---", config)
+	cfg := pluginConfig{}
+	err := yaml.Unmarshal(config, &cfg)
+	if err != nil {
+		errWithContext := fmt.Errorf("setConfig: cannot parse configuration: %w", err)
+		log.Debugf("%s", errWithContext)
+		return errWithContext
+	}
+	p.config = &cfg
+	log.Tracef("new configuration has %d classes:", len(p.config.Classes))
+	for _, cls := range p.config.Classes {
+		log.Tracef("- %s", cls.Name)
+	}
+	return nil
+}
+
+// pprintCtr() returns human readable container name that is
+// unique to the node.
+func pprintCtr(pod *api.PodSandbox, ctr *api.Container) string {
+	return fmt.Sprintf("%s/%s:%s", pod.GetNamespace(), pod.GetName(), ctr.GetName())
+}
+
+// applyQosClass applies QoS class to a container, updates unified values.
+func (p *plugin) applyQosClass(pod *api.PodSandbox, ctr *api.Container, cls string, unifiedp *map[string]string) error {
+	if p.config == nil {
+		return fmt.Errorf("missing plugin configuration")
+	}
+	for _, class := range p.config.Classes {
+		log.Tracef("comparing configuration class %q to annotation %q", class.Name, cls)
+		if class.Name == cls {
+			log.Tracef("applying SwapLimitRatio=%.2f on unified=%v", class.SwapLimitRatio, *unifiedp)
+			if class.SwapLimitRatio > 0 {
+				memLimitp := ctr.Linux.Resources.Memory.Limit
+				if memLimitp == nil {
+					return fmt.Errorf("missing container memory limit")
+				}
+				// memory.high and memory.swap.max
+				// values defined by the QoS class do
+				// not override these values if set by
+				// specifically with unified annotations.
+				associate(unifiedp, "memory.high", strconv.FormatInt(int64(float32(memLimitp.Value)*(1.0-class.SwapLimitRatio)), 10), false)
+				associate(unifiedp, "memory.swap.max", "max", false)
+			}
+			log.Tracef("resulted unified=%v", *unifiedp)
+			return nil
+		}
+	}
+	return fmt.Errorf("class not found in plugin configuration")
+}
+
+// associate adds new key-value pair to a map, or updates existing
+// pair if called with override. Returns true if added/updated.
+func associate(m *map[string]string, key, value string, override bool) bool {
```

Comment:
same nitpick as in #117: why to use a pointer here ?

Reply:
Fixed